ARCHITECTURE.md

```diff
@@ -0,0 +1,184 @@
+  tests. Note that some libraries do not have mocks.
+- `quickstart/` contains a small program showing how to use the library in a
+  larger project. This is intended for application developers wanting to, well,
+  use the libraries quickly.
+- `samples/` (sometimes `examples/`) contain code snippets that are extracted
+  both into the Doxygen reference documentation and to the cloud.google.com
+  site, as part of the service documentation. Some libraries have a single
+  program in this directory, others have smaller programs (or more samples).
+
+The remaining directories are implementation details, or at least only intended
+to be interesting for `google-cloud-cpp` developers and contributors.
+
+- `internal/` contains implementation details for the library. The files in
+  these directories are *not* intended for public consumption, but contributors
+  and developers need to change them, of course. The relationship between `.h`,
+  `.cc` and `_test.cc` files mirrors the one found in `${library}/`.
+- `benchmarks/` contains "macro benchmarks" for the library, typically programs
+  that interact with the production environment to verify the library meets the
+  latency, throughput, and CPU overhead (as in "low CPU overhead") requirements.
+- `ci/` may contain helper scripts used in the CI builds, for example, to
+  startup and shutdown any emulators used by the library.
+- `doc/` contains the landing page for any Doxygen documentation, as well as
+  additional Doxygen docs that are not extracted from the code comments.
+- `integration_tests/` (or simply `tests/` in some libraries) contain any
+  integration tests where we test the library using emulators and/or the
+  production environment.
+- `testing/` contains testing utilities shared by the unit tests and the
+  integration tests of the library. They are not intended to be part of the
+
+## The `*Client` classes
+
+The main interface for each library is the `${library}::*Client` class. Roughly
+speaking each "client" corresponds to a `service` in the `.proto` definitions
+for the service. It is common for services to have separate `service`
+definitions for "admin" operations, such as creating a new instance of the
+service, or changing permissions, as opposed to "normal" operations, such as
+inserting a new row, or publishing a new message. When "admin" services are
+defined, there are separate `${library}::Client` objects.
+
+These are some examples:
+
+- Bigtable has a [bigtable::Table](/google/cloud/bigtable/table.h) client, as
+  well as [TableAdmin](/google/cloud/bigtable/table_admin.h) and
+  [InstanceAdmin](/google/cloud/bigtable/instance_admin.h) clients.
+- Storage has a single [storage::Client](/google/cloud/storage/client.h) client
+  class.
+- Pub/Sub has two non-admin clients, the
+  [pubsub::Publisher](/google/cloud/pubsub/publisher.h)
+  and the [pubsub::Subscriber](/google/cloud/pubsub/subscriber.h).
+- Spanner has one non-admin client, the
```

Comment:
Spanner also has the `DatabaseAdminClient` and `InstanceAdminClient`. Did you want to mention those here too?

Reply:
Done.